### PR TITLE
Issue #8018: OperatorWrapCheck no violation if single line comment 

### PR DESCRIPTION
**Pull request by harsh-kukreja**
Fix for #8018 

DIff report: 
https://harsh-kukreja.github.io/issue-8018/diff/

### CODE REVIEW

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java [hunk 1]
```diff
@@ -308,4 +316,9 @@ else if (option == WrapOption.EOL
         }
     }
 
+    @Override
+    public boolean isCommentNodesRequired() {
+        return true;
```
Comment (rnveach) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java:321]: Why is this needed? If this is needed, check may become more complicated to validate as now we need to determine if comments showing up in the AST tree is causing a separate issue.

Comment (harsh-kukreja) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java:321]: If we donot set this true then there will be no comment nodes found so how will i check where the tree contains a comment node?

**Comment by harsh-kukreja** [comment]
> This change will need regression. 
Okay i will update the pr once the jococo report is final @rnveach 
can u help me with how to run the command for report?

**Comment by rnveach** [comment]
https://github.com/checkstyle/checkstyle/wiki/How-to-run-certain-phases-and-validations

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java (hunk 1 above)
Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java:321]: Do you need to know more about comments more than reported issue, which is a line ending with a comment? Since we are reading the source why can't we check for `//` or `/*`? Will we have further issues with `/*` since more code can be after `*/`?

If we need to add this method, we will need to split the work into 2 commits and provide regression between each one.

1st commit should be adding this method and any support now that this method is added and regression should expect no differences.
2nd commit is to fix the original issue and regression should show differences.

Comment (harsh-kukreja) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java:320]: Okay so I will just check for `//` and `/*` in the lines @rnveach this will be much better

Comment (rnveach) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java:320]: @harsh-kukreja If any good violations are removed from regression because of incorrect handling of  `/*`, we might have to fix it.

**Comment by harsh-kukreja** [comment]
Hey @rnveach can you tell me how to solve this 
https://travis-ci.org/github/checkstyle/checkstyle/jobs/674668350#L1239

**Comment by rnveach** [comment]
> looking at the diff report there are some violations in the checkstyle repo 

CI tells us if this PR needs additional fixes for our code and style. Regression is showing new violations in test resource files which are not as strictly enforced on style as production code. They are most likely "bad" because they are testing something specific.

CI is passing in this regard so you don't need any additional fixes.

> I don't know how to update the pmd files as the ci is failing so please help

https://travis-ci.org/github/checkstyle/checkstyle/jobs/674668350
````
[INFO] There are 3 errors reported by Checkstyle 8.32-SNAPSHOT with /net/sourceforge/pmd/pmd-checkstyle-config.xml ruleset.
[ERROR] src/main/java/net/sourceforge/pmd/cpd/AbstractTokenizer.java:[120,19] (whitespace) OperatorWrap: '&&' should be on a new line.
[ERROR] src/main/java/net/sourceforge/pmd/cpd/AbstractTokenizer.java:[121,45] (whitespace) OperatorWrap: '&&' should be on a new line.
[ERROR] src/main/java/net/sourceforge/pmd/cpd/AbstractTokenizer.java:[125,41] (whitespace) OperatorWrap: '&&' should be on a new line.
````

PMD is showing new violations from regression. They must be reviewed and if they are correct, a PR must be sent to them to correct the code for the CI to pass and eventually be merged.

Their repo is located at https://github.com/checkstyle/checkstyle/blob/master/.ci/travis/travis.sh#L457 ( https://github.com/pmd/pmd.git )

> can you tell me how to solve this 

Please ask a direct question if I missed anything else.

**Comment by harsh-kukreja** [comment]
Hey @rnveach I checked the pmd file the violations are correct according to me you can have a look 
https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/AbstractTokenizer.java#L120

https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/AbstractTokenizer.java#L121

https://github.com/pmd/pmd/blob/master/pmd-core/src/main/java/net/sourceforge/pmd/cpd/AbstractTokenizer.java#L125

So now I should open an issue or directly solve and open up a pr can you tell me on how should i proceed?

**Comment by harsh-kukreja** [comment]
Hey @rnveach the fix is done and merge can u restart travis 
you can view it here https://github.com/pmd/pmd/pull/2423

**Comment by harsh-kukreja** [comment]
Hey @rnveach rply

**Comment by rnveach** [comment]
@harsh-kukreja CI still failed on a re-run in a different area.
````
[INFO] --- maven-checkstyle-plugin:3.1.1:check (checkstyle-check) @ pmd-java ---
[INFO] There is 1 error reported by Checkstyle 8.32-SNAPSHOT with /net/sourceforge/pmd/pmd-checkstyle-config.xml ruleset.
[ERROR] src/main/java/net/sourceforge/pmd/lang/java/rule/errorprone/CloneMethodMustImplementCloneableRule.java:[94,38] (whitespace) OperatorWrap: '&&' should be on a new line.
````

Please run your check on the entire source. CI will stop at just the first error it finds.

**Comment by harsh-kukreja** [comment]
Hey @rnveach please restart the travis

**Comment by strkkk** [comment]
I have restarted travis

**Comment by harsh-kukreja** [comment]
Hey @rnveach please review

**Comment by pbludov** [comment]
@harsh-kukreja please explain this violation:
https://harsh-kukreja.github.io/issue-8018/diff/spotbugs/xref/Users/Harsh/Desktop/contribution/checkstyle-tester/repositories/spotbugs/spotbugs/src/main/java/edu/umd/cs/findbugs/ba/jsr305/TypeQualifierValueSet.java.html#L335

It looks like a false positive for me.

**Comment by harsh-kukreja** [comment]
Done @pbludov added an example also in the input file. Sorry i missed that case
Do i need to generate the diff report again?

**Comment by pbludov** [comment]
> Do i need to generate the diff report again?

Yes, we must be sure that these violations have disappeared.

**Comment by harsh-kukreja** [comment]
Hey @pbludov updated the diff report

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java [hunk 2]
```diff
@@ -309,4 +310,26 @@ else if (option == WrapOption.EOL
...
+    /**
+     * Checks for multiline comment start and end. Also checks if there is ant code after comment.
+     *
+     * @param currentLineSubstring A string to check.
+     * @return true If it ends and there is no code after the end,
+     *     If it doesn't end on the same line
+     * */
+    private static boolean checkCodeForMultilineComment(String currentLineSubstring) {
+        boolean result = false;
+        if (currentLineSubstring.trim().indexOf("/*") == 0) {
```
Comment (pbludov) [inline src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java:322]: There is tree calls to `currentLineSubstring.trim()` in this method (and one in `visitToken`).
Could you please refactor the code to call it once? Also, `CommonUtil.isBlank(currentLineSubstring)` is same as `currentLineSubstring.trim().isEmpty()`. Please move all new code to this method, so `visitToken` will be something like 
```java
 if (option == WrapOption.NL
                    && !text.equals(currentLine.trim())
                    && containsNoCode(currentLineSubstring.trim()))

```
True if substring is blank;
starts with // comment;
starts with /* and ends with */
`boolean containsNoCode(String substring)`

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java [hunk 3]
```diff
@@ -309,4 +310,26 @@ else if (option == WrapOption.EOL
         }
     }
 
+    /**
+     * Checks for multiline comment start and end. Also checks if there is ant code after comment.
```
[the remark this answers is not in the document]
Comment (harsh-kukreja) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java]: Done @pbludov

File: src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java (hunk 2 above)
Comment (harsh-kukreja) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java]: Done @pbludov

Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java]: @harsh-kukreja I am sorry for making you do extra work, but please pay attention to https://github.com/checkstyle/checkstyle/issues/8018#issuecomment-621872177
If it is possible to use the AST to check for violations, then we should use it.
Please investigate.

Comment (harsh-kukreja) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java]: Hey @pbludov I gave u an rply there itself please check

Comment (pbludov) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java]: @harsh-kukreja here is my AST based solution:
```
    @Override
    public void visitToken(DetailAST ast) {
        final DetailAST parent = ast.getParent();
        // we do not want to check colon for cases and defaults
        if (parent.getType() != TokenTypes.LITERAL_DEFAULT
                && parent.getType() != TokenTypes.LITERAL_CASE) {
            if (option == WrapOption.NL
                    && TokenUtil.areOnSameLine(ast, getFirstOperand(ast))
                    && !TokenUtil.areOnSameLine(ast, getSecondOperand(ast))) {
                log(ast, MSG_LINE_NEW, ast.getText());
            }
            else if (option == WrapOption.EOL && !TokenUtil.areOnSameLine(ast, getFirstOperand(ast))) {
                log(ast, MSG_LINE_PREVIOUS, ast.getText());
            }
        }
    }

    private static DetailAST getFirstOperand(DetailAST ast) {
        DetailAST result;
        if (ast.getFirstChild() != null && ast.getParent().getType() != TokenTypes.VARIABLE_DEF) {
            result = ast.getFirstChild();
            while (result.getLastChild() != null) {
                result = result.getLastChild();
            }
        }
        else {
            result = ast.getPreviousSibling();
        }
        return result;
    }

    private static DetailAST getSecondOperand(DetailAST ast) {
        DetailAST result;
        if (ast.getLastChild() != null) {
            result = ast.getLastChild();
            while (result.getFirstChild() != null) {
                result = result.getFirstChild();
            }
        }
        else {
            result = ast.getNextSibling();
        }
        return result;
    }
```

It may be incomplete, there may be more nuances.
For now, it handles only one of them: `ASSIGN` in `VARIABLE_DEF` has only one child:
```
        int x = 0;
        x = x + 2;
```
The AST:
```
|--VARIABLE_DEF -> VARIABLE_DEF [5:8]
|   |--MODIFIERS -> MODIFIERS [5:8]
|   |--TYPE -> TYPE [5:8]
|   |   `--LITERAL_INT -> int [5:8]
|   |--IDENT -> x [5:12]
|   `--ASSIGN -> = [5:14]              // Left operand is previous sibling
|       `--EXPR -> EXPR [5:16]
|           `--NUM_INT -> 0 [5:16]
|--SEMI -> ; [5:17]
|--EXPR -> EXPR [6:10]
|   `--ASSIGN -> = [6:10]              // Left operand is first child
|       |--IDENT -> x [6:8]
|       `--PLUS -> + [6:14]
|           |--IDENT -> x [6:12]
|           `--NUM_INT -> 2 [6:16]
|--SEMI -> ; [6:17]
```
Please use this code as an example and fix all cases discovered by regression.

Comment (harsh-kukreja) [reply src/main/java/com/puppycrawl/tools/checkstyle/checks/whitespace/OperatorWrapCheck.java]: Hey @pbludov if we add the comment nodes there will be more complication in the check and the code u wrote is complicated than the code currently it will be great if we go forward with the code i wrote in the Pull Request. @rnveach your views about this

**Comment by romani** [comment]
@pbludov , do you have time to finish this PR ?

**Comment by pbludov** [comment]
> @pbludov , do you have time to finish this PR ?

I'm on it

**Comment by pbludov** [comment]
@harsh-kukreja the work will continue in PR #9008
This PR is closed.
